### DISCUSSION TITLE
Code review / Q&A request - colored wireframes

### DISCUSSION THREAD

**Discussion by LightAndLight** [bevyengine/bevy#6358] [Q&A]
Bevy's [current wireframe plugin](https://github.com/bevyengine/bevy/blob/3c13c75036108eec91beba58621e32e50c1be11d/crates/bevy_pbr/src/wireframe.rs) hard-codes the wireframe color. I wanted per-entity wireframe colors, so I created my own plugin based on the one in `bevy_pbr`.

I've got it working, but suspect there's a nicer way to do it. There are also a few Bevy renderer things that I copied from the original plugin that I don't understand.

If you're feeling generous with your time, I'd love to hear your feedback. I also have some specific questions, in case you don't want to peruse my code.

My plugin:

* [colored_wireframe.rs](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs)
* [wireframe.wgsl](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl)

Specific questions:

* [colored_wireframe.rs#L61-L75](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L61-L75) - Is there a better place to create and store the bind group layout?

* [colored_wireframe.rs#L80-L103](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L80-L103), [colored_wireframe.rs#L219](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L219), [colored_wireframe.rs#L233](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L233), [colored_wireframe.rs#L241-L257](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L241-L257), [colored_wireframe.rs#L296](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L296) - What is pipeline specialization and why do I need to use it?
  * Since I didn't set up imports ([wireframe.wgsl#L1-L2](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L1-L2), [wireframe.wgsl#L10](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L10), [wireframe.wgsl#L17](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L17)), preprocessor variables ([wireframe.wgsl#L7](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L7)), vertex input locations ([wireframe.wgsl#L19-L25](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L19-L25)), or the first two bind groups ([wireframe.wgsl#L4-L5](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L4-L5), [wireframe.wgsl#L8-L9](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/render/wireframe.wgsl#L8-L9)), I assume this are defined in the mesh pipeline and "inherited" by my custom pipeline when I specialize. I think I can figure out how it all works by reading the code. I'd appreciate a prose explanation to prime me first, though.
  * Related resources:
    * Pipeline specialization - https://bevyengine.org/news/bevy-0-6/#pipeline-specialization
    * Mesh pipeline specialization - https://bevyengine.org/news/bevy-0-7/#flexible-mesh-vertex-layouts

* [colored_wireframe.rs#L94-L99](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L94-L99) - This feels janky. Is there a better way to add a bind group layout to the layout I inherited from the mesh pipeline?

* [colored_wireframe.rs#L154-L172](https://github.com/LightAndLight/hover-pill/blob/colored-wireframes-code-review/src/colored_wireframe.rs#L154-L172) - Is it expensive to re-create a bind group each frame? The `DynamicUniformBuffer` I'm using might allocate new graphics buffers without me knowing, and binding every frame was the easiest way to make things work.

**Comment by LegNeato** [comment]
I'm not a member of the project, but it is likely better to just put it up as a draft PR.
